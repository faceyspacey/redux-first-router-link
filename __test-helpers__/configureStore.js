import { createStore, applyMiddleware, compose, combineReducers } from 'redux'
import { connectRoutes } from 'redux-first-router'
import thunk from 'redux-thunk'

const configureStore = (initialPath, options, additionalReducers) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param'
  }

  const { middleware, enhancer, reducer } = connectRoutes(routesMap, {
    initialEntries: [initialPath || '/'],
    initialIndex: 0,
    keyLength: 6,
    ...options
  })

  const middlewares = applyMiddleware(middleware, thunk)
  const enhancers = compose(enhancer, middlewares)
  const rootReducer = combineReducers({
    ...additionalReducers,
    location: reducer
  })

  return createStore(rootReducer, enhancers)
}

export default configureStore
