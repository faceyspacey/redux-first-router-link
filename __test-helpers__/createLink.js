import React from 'react'
import renderer from 'react-test-renderer'
import { createStore, applyMiddleware, compose } from 'redux'
import { Provider } from 'react-redux'
import { connectRoutes } from 'redux-first-router'

import Link from '../src/Link'
import NavLink from '../src/NavLink'

const createLink = (props, initialPath, options) => {
  const isNavLink = !!initialPath
  const link = isNavLink ? <NavLink {...props} /> : <Link {...props} />

  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param'
  }

  const { middleware, enhancer, reducer } = connectRoutes(routesMap, {
    initialEntries: [initialPath || '/'],
    initialIndex: 0,
    keyLength: 6,
    ...options
  })

  const middlewares = applyMiddleware(middleware)
  const enhancers = compose(
    enhancer,
    middlewares
  )
  const rootReducer = (state = {}, action = {}) => ({
    location: reducer(state.location, action)
  })

  const store = createStore(rootReducer, enhancers)
  const component = renderer.create(
    <Provider store={store}>{link}</Provider>,
    { createNodeMock: () => 'node' },
  )

  return {
    component,
    tree: component.toJSON(),
    store
  }
}

export default createLink
export const createNavLink = (path, props) => createLink(props, path)

export const event = { preventDefault: () => undefined, button: 0 }
