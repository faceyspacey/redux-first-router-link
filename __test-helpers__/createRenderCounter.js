import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import configureStore from './configureStore'

const unrelated = (state = 0, { type }) =>
  type === 'INC_UNRELATED' ? state + 1 : state

const createRenderCounter = (connector, countWrapper) => {
  const store = configureStore(undefined, undefined, { unrelated })

  const RenderCounter = connector(() => {
    ++countWrapper.count
    return <span />
  })

  const component = renderer.create(
    <Provider store={store}>
      <RenderCounter />
    </Provider>
  )

  return {
    component,
    store
  }
}

export default createRenderCounter
