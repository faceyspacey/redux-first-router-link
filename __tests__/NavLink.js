import { NOT_FOUND } from 'redux-first-router'
import { createNavLink, event } from '../__test-helpers__/createLink'

test('NON-EXACT: show active class', () => {
  const { tree } = createNavLink('/first', {
    to: '/first',
    activeClassName: 'active'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */
})

test('EXACT: DONT show active class', () => {
  const { tree } = createNavLink('/second/dog', {
    to: '/second',
    exact: true,
    activeClassName: 'active'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */
})

test('STRICT: DONT show active class', () => {
  const { tree } = createNavLink('/first', {
    to: '/first/',
    strict: true,
    activeClassName: 'active'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */
})

test('show activeStyle', () => {
  const { tree } = createNavLink('/first', {
    to: '/first',
    activeStyle: { color: 'red' }
  }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */
})

test('combine with existing styles and class', () => {
  const { tree } = createNavLink('/first', {
    to: '/first',
    className: 'foo',
    style: { fontSize: 32 },
    activeClassName: 'active',
    activeStyle: { color: 'red' }
  }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */
})

test('reacts to state changes (onClick)', () => {
  const { tree, store, component } = createNavLink('/first', {
    to: '/second/bar',
    activeClassName: 'active'
  })

  expect(tree.props.className).not.toBeDefined()
  expect(tree).toMatchSnapshot() /*? tree */

  // store.dispatch({ type: 'SECOND', payload: { param: 'bar' } })
  tree.props.onClick(event) // this dispatches above action (obviously)

  const tree2 = component.toJSON() /*? */

  expect(tree2.props.className).toEqual('active')
  expect(tree2).toMatchSnapshot()
})

test('isActive returns true', () => {
  const { tree, store } = createNavLink('/first', {
    to: '/first',
    activeClassName: 'active',
    isActive: (match, location) => {
      expect(match).toMatchSnapshot()
      expect(location).toMatchSnapshot()
      return match
    }
  })

  expect(tree).toMatchSnapshot()
})

test('isActive return false', () => {
  const { tree, store } = createNavLink('/first', {
    to: '/first',
    activeClassName: 'active',
    isActive: (match, location) => {
      expect(match).toMatchSnapshot()
      expect(location).toMatchSnapshot()
      return false
    }
  })

  expect(tree).toMatchSnapshot()
})

it('supports custom HTML tag name', () => {
  const { tree, store } = createNavLink('/first', {
    to: '/second',
    activeClassName: 'active',
    tagName: 'div'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot()
})

it('supports custom HTML tag name in active mode', () => {
  const { tree, store } = createNavLink('/first', {
    to: '/first',
    activeClassName: 'active-foo',
    tagName: 'div'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot()
})

it('supports custom HTML tag name which is still a link', () => {
  const { tree, store } = createNavLink('/first', {
    to: 'somewhere',
    tagName: 'a'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot()
})

test('query params are ommitted', () => {
  const { tree } = createNavLink('/first', {
    to: '/first?foo=123',
    activeClassName: 'active'
  })

  expect(tree).toMatchSnapshot()
})

test('function as Child is executed with active boolean', () => {
  const children = jest.fn(() => '')
  const { tree } = createNavLink('/first', {
    to: '/first',
    children
  })

  expect(children).toBeCalledWith(true)
})
