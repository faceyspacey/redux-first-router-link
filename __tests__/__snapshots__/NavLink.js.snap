// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EXACT: DONT show active class 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="/second"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`NON-EXACT: show active class 1`] = `
<a
  aria-current="true"
  className="active"
  href="/first"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`STRICT: DONT show active class 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="/first/"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`combine with existing styles and class 1`] = `
<a
  aria-current="true"
  className="foo active"
  href="/first"
  onClick={[Function]}
  style={
    Object {
      "color": "red",
      "fontSize": 32,
    }
  }
/>
`;

exports[`isActive return false 1`] = `
Object {
  "isExact": true,
  "params": Object {},
  "path": "/first",
  "url": "/first",
}
`;

exports[`isActive return false 2`] = `
Object {
  "hasSSR": undefined,
  "history": undefined,
  "kind": "load",
  "pathname": "/first",
  "payload": Object {},
  "prev": Object {
    "pathname": "",
    "payload": Object {},
    "type": "",
  },
  "routesMap": Object {
    "FIRST": "/first",
    "HOME": "/",
    "SECOND": "/second/:param",
  },
  "type": "FIRST",
}
`;

exports[`isActive return false 3`] = `
<a
  aria-current={false}
  className={undefined}
  href="/first"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`isActive returns true 1`] = `
Object {
  "isExact": true,
  "params": Object {},
  "path": "/first",
  "url": "/first",
}
`;

exports[`isActive returns true 2`] = `
Object {
  "hasSSR": undefined,
  "history": undefined,
  "kind": "load",
  "pathname": "/first",
  "payload": Object {},
  "prev": Object {
    "pathname": "",
    "payload": Object {},
    "type": "",
  },
  "routesMap": Object {
    "FIRST": "/first",
    "HOME": "/",
    "SECOND": "/second/:param",
  },
  "type": "FIRST",
}
`;

exports[`isActive returns true 3`] = `
<a
  aria-current="true"
  className="active"
  href="/first"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`query params are ommitted 1`] = `
<a
  aria-current="true"
  className="active"
  href="/first?foo=123"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`reacts to state changes (onClick) 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="/second/bar"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`reacts to state changes (onClick) 2`] = `
<a
  aria-current="true"
  className="active"
  href="/second/bar"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`show activeStyle 1`] = `
<a
  aria-current="true"
  className="active"
  href="/first"
  onClick={[Function]}
  style={
    Object {
      "color": "red",
    }
  }
/>
`;

exports[`supports custom HTML tag name 1`] = `
<div
  aria-current={false}
  className={undefined}
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`supports custom HTML tag name in active mode 1`] = `
<div
  aria-current="true"
  className="active-foo"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`supports custom HTML tag name which is still a link 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="somewhere"
  onClick={[Function]}
  style={undefined}
/>
`;
